execution-engine/engine-core/src/execution/runtime/mod.rs

```diff
@@ -560,6 +575,577 @@ where
         Ok(result)
     }
 
+    #[allow(clippy::cognitive_complexity)]
+    fn extract_urefs(value: &CLValue) -> Result<Vec<URef>, Error> {
+        match value.cl_type() {
+            CLType::Bool
+            | CLType::I32
+            | CLType::I64
+            | CLType::U8
+            | CLType::U32
+            | CLType::U64
+            | CLType::U128
+            | CLType::U256
+            | CLType::U512
+            | CLType::Unit
+            | CLType::String
+            | CLType::Any => Ok(vec![]),
+            CLType::Option(ty) => match **ty {
+                CLType::URef => {
+                    let opt: Option<URef> = value.to_owned().into_t()?;
+                    Ok(opt.into_iter().collect())
+                }
+                CLType::Key => {
+                    let opt: Option<Key> = value.to_owned().into_t()?;
+                    Ok(opt.into_iter().flat_map(Key::into_uref).collect())
+                }
+                _ => Ok(vec![]),
+            },
+            CLType::List(ty) => match **ty {
+                CLType::URef => Ok(value.to_owned().into_t()?),
+                CLType::Key => {
+                    let keys: Vec<Key> = value.to_owned().into_t()?;
+                    Ok(keys.into_iter().filter_map(Key::into_uref).collect())
+                }
+                _ => Ok(vec![]),
+            },
+            CLType::FixedList(ty, 1) => match **ty {
```

Comment:
For the `FixedList`, I was thinking we could do something like the following:
```rust
CLType::FixedList(ty, length) => match **ty {
    CLType::URef => {
        let mut urefs = Vec::with_capacity(*length as usize);
        let mut bytes = value.inner_bytes().as_slice();
        for _ in 0..*length {
            let (uref, remainder) = URef::from_bytes(bytes)?;
            urefs.push(uref);
            bytes = remainder;
        }
        Ok(urefs)
    }
    CLType::Key => {
        let mut urefs = Vec::with_capacity(*length as usize);
        let mut bytes = value.inner_bytes().as_slice();
        for _ in 0..*length {
            let (key, remainder) = Key::from_bytes(bytes)?;
            if let Some(uref) = key.as_uref() {
                urefs.push(*uref);
            }
            bytes = remainder;
        }
        Ok(urefs)
    }
    _ => Ok(vec![]),
},
```
We'd need to change the default impls in the `impl_to_from_bytes_for_array` macro to not prepend the length of the array.  That's already the case for the specialisation of arrays of `u8`.  (I actually thought we already avoided prepending lengths to arrays.)